File: python/src/facebook_engagement_mcp/graph/normalize.py
from __future__ import annotations

from typing import Any

from .types import Author, Comment, PageCredential, PagePost


def normalize_comment(raw: dict[str, Any]) -> Comment:
    author = raw.get("from")
    return Comment(
        id=raw["id"],
        message=raw.get("message"),
        created_time=raw.get("created_time"),
        author=Author(id=author.get("id"), name=author.get("name")) if author else None,
        like_count=raw.get("like_count"),
        reply_count=raw.get("comment_count"),
        hidden=raw.get("is_hidden"),
        can_comment=raw.get("can_comment"),
        can_hide=raw.get("can_hide"),
        permalink=raw.get("permalink_url"),
        parent_id=(raw.get("parent") or {}).get("id"),
    )


def normalize_post(raw: dict[str, Any]) -> PagePost:
    return PagePost(
        id=raw["id"],
        message=raw.get("message"),
        created_time=raw.get("created_time"),
        permalink=raw.get("permalink_url"),
        # Defaulting to True matches the TypeScript client. `/feed` never
        # returns an unpublished post at all (T-02), so this default describes
        # what the edge can actually contain.
        is_published=raw.get("is_published", True),
    )


def normalize_credential(raw: dict[str, Any]) -> PageCredential:
    return PageCredential(
        id=raw["id"],
        name=raw.get("name"),
        access_token=raw.get("access_token"),
        tasks=list(raw.get("tasks") or []),
    )

File: go/internal/graph/normalize.go
package graph

// Raw Graph shapes to the shapes the rest of this server speaks.
//
// Every conversion here preserves absence. A pointer that arrives nil leaves
// nil; it is never replaced with a zero value, because downstream the two mean
// different things. The single exception is PagePost.IsPublished, and it is
// commented where it happens.

func NormalizePage(raw RawPage) Page {
	return Page{
		ID:       raw.ID,
		Name:     raw.Name,
		Category: raw.Category,
		Tasks:    raw.Tasks,
	}
}

func NormalizePost(raw RawPost) PagePost {
	// The one defaulting field. Graph omits is_published on some edges, and an
	// omitted value must not be read as "this is an ad post" — that would label
	// ordinary organic posts as ad-backed throughout the response.
	published := true
	if raw.IsPublished != nil {
		published = *raw.IsPublished
	}

	return PagePost{
		ID:          raw.ID,
		Message:     raw.Message,
		CreatedTime: raw.CreatedTime,
		Permalink:   raw.PermalinkURL,
		IsPublished: published,
	}
}

func NormalizePhoto(raw RawPhoto) PagePhoto {
	return PagePhoto{
		ID:          raw.ID,
		CreatedTime: raw.CreatedTime,
		PostID:      raw.PageStoryID,
	}
}

func NormalizeComment(raw RawComment) Comment {
	comment := Comment{
		ID:          raw.ID,
		Message:     raw.Message,
		CreatedTime: raw.CreatedTime,
		LikeCount:   raw.LikeCount,
		ReplyCount:  raw.CommentCount,
		Hidden:      raw.IsHidden,
		CanComment:  raw.CanComment,
		CanHide:     raw.CanHide,
		Permalink:   raw.PermalinkURL,
	}

	// `from` present with no id is not an identity, but it IS a returned field.
	// Keep the distinction rather than collapsing it: nil means Graph said
	// nothing, and a non-nil Author with a nil ID means Graph said something
	// that does not identify anyone.
	if raw.From != nil {
		comment.Author = &Author{ID: raw.From.ID, Name: raw.From.Name}
	}

	// A parent object with no id does not invent one.
	if raw.Parent != nil && raw.Parent.ID != nil {
		comment.ParentID = raw.Parent.ID
	}

	return comment
}
